0013-RS-pubsub-subscription-streaming.md

```diff
@@ -0,0 +1,195 @@
+
+## Expectations and alternatives
+
+This proposal outlines the gRPC & HTTP streaming API for subscribing to PubSub topics.
+This proposal does _not_ address any hot-reloading functionality to the existing programmatic or declarative subscription configuration.
+Using a gRPC streaming API is the most natural fit for this feature, as it allows for first class long-lived bi-directional connections to Dapr to receive messages.
+A supplementary WebSocket based HTTP API is useful for applications which do not have a gRPC client available or HTTP WebSockets are preferred.
+These messages are typed RPC giving the best UX in each SDK.
+Once implemented, this feature will need to be implemented in all Dapr SDKs.
+
+## Solution
+
+### gRPC
+
+Rough gRPC PoC implementation: https://github.com/dapr/dapr/commit/ed40c95d11b78ab9a36a4a8f755cf89336ae5a05
+
+The Dapr runtime gRPC server will implement the following new RPC and messages:
+
+```proto
+service Dapr {
+  // SubscribeTopicEventsAlpha1 subscribes to a PubSub topic and receives topic events
+  // from it.
+  rpc SubscribeTopicEventsAlpha1(stream SubscribeTopicEventsRequestAlpha1) returns (stream TopicEventRequestAlpha1) {}
+}
+
+// SubscribeTopicEventsRequest is a message containing the details for
+// subscribing to a topic via streaming.
+// The first message must always be the initial request. All subsequent
+// messages must be event responses.
+message SubscribeTopicEventsRequestAlpha1 {
+  oneof subscribe_topic_events_request_type {
+    SubscribeTopicEventsSubscribeRequestAlpha1 request = 1;
+    SubscribeTopicEventsResponseAlpha1 event_response = 2;
+  }
+}
+
+// SubscribeTopicEventsSubscribeRequest is the initial message containing the
+// details for subscribing to a topic via streaming.
+message SubscribeTopicEventsSubscribeRequestAlpha1 {
+  // The name of the pubsub component
+  string pubsub_name = 1 [json_name = "pubsubName"];
+
+  // The pubsub topic
+  string topic = 2 [json_name = "topic"];
+
+  // The metadata passing to pub components
+  //
+  // metadata property:
+  // - key : the key of the message.
+  optional map<string, string> metadata = 3 [json_name = "metadata"];
+
+  // dead_letter_topic is the topic to which messages that fail to be processed
+  // are sent.
+  optional string dead_letter_topic = 4 [json_name = "deadLetterTopic"];
+
+  // max_in_flight_messages is the maximum number of in-flight messages that
+  // can be processed by the subscriber at any given time.
+  // Default is no limit.
+  optional max_in_flight_messages = 5 [json_name = "maxInFlightMessages"];
+}
+
+// SubscribeTopicEventsResponse is a message containing the result of a
+// subscription to a topic.
+message SubscribeTopicEventsResponseAlpha1 {
+  // id is the unique identifier for the subscription request.
+  string id = 1 [json_name = "id"];
+
+  // status is the result of the subscription request.
+  TopicEventResponseAlpha1 status = 2 [json_name = "status"];
```

Comment:
`TopicEventResponse` refers to a response on a [published message](https://github.com/dapr/dapr/blob/b1959bd51621885ccf6bc4763093b91977f4be7f/dapr/proto/runtime/v1/appcallback.proto#L106) am I right in reading this as it being used as a response for the subscription response too? If so seems a bit too tightly coupled in my opinion